edb/server/protocol/auth/scram.py

```diff
@@ -0,0 +1,310 @@
+            )
+            response.status = http.HTTPStatus.BAD_REQUEST
+            response.close_connection = True
+            return
+
+        if (
+            not scram.verify_client_proof(
+                client_first_bare,
+                server_first,
+                client_final_without_proof,
+                verifier.stored_key,
+                client_proof,
+            )
+            or mock_auth
+        ):
+            response.body = b"Authentication failed"
+            response.status = http.HTTPStatus.UNAUTHORIZED
+            response.custom_headers["WWW-Authenticate"] = "SCRAM-SHA-256"
+            return
+
+        server_final = base64.b64encode(
+            scram.build_server_final_message(
+                client_first_bare,
+                server_first,
+                client_final_without_proof,
+                verifier.server_key,
+            ).encode("utf-8")
+        ).decode("ascii")
+
+        try:
+            response.body = generate_jwt_token(username, server)
+        except ValueError as ex:
+            if debug.flags.server:
+                markup.dump(ex)
+            response.body = b"Authentication failed"
+            response.status = http.HTTPStatus.UNAUTHORIZED
+            response.custom_headers["WWW-Authenticate"] = "SCRAM-SHA-256"
+            return
+
+        response.custom_headers[
+            "Authentication-Info"
+        ] = f"sid={sid}, data={server_final}"
+
+
+def get_scram_verifier(user, server):
+    roles = server.get_roles()
+
+    rolerec = roles.get(user)
+    if rolerec is not None:
+        verifier_string = rolerec["password"]
+        if verifier_string is None:
+            raise ValueError(f"invalid SCRAM verifier for user {user!r}")
+
+        verifier = scram.parse_verifier(verifier_string)
+        is_mock = False
+    else:
+        # To avoid revealing the validity of the submitted user name,
+        # generate a mock verifier using a salt derived from the
+        # received user name and the cluster mock auth nonce.
+        # The same approach is taken by Postgres.
+        nonce = server.get_instance_data("mock_auth_nonce")
+        salt = hashlib.sha256(nonce.encode() + user.encode()).digest()
+
+        verifier = scram.SCRAMVerifier(
+            mechanism="SCRAM-SHA-256",
+            iterations=scram.DEFAULT_ITERATIONS,
+            salt=salt[: scram.DEFAULT_SALT_LENGTH],
+            stored_key=b"",
+            server_key=b"",
+        )
+
+        is_mock = True
+
+    return verifier, is_mock
+
+
+def generate_jwt_token(user, server):
+    rolerec = server.get_roles().get(user)
+    if rolerec is None or rolerec["password"] is None:
+        raise ValueError(f"invalid user: {user!r}")
+    ekey = server.get_jwe_key()
+    skey = server.get_jws_key()
+
+    namespace = "edgedb.server"
+    token = jwt.JWT(
+        header={"alg": "ES256" if skey["kty"] == "EC" else "RS256"},
+        claims={f"{namespace}.roles": {user: rolerec["password"]}},
```

Comment:
Let's also include the standard `iat` (issued at) claim to control for token expiration.  The latter would probably be configurable somehow, but let's worry about that later.